core/src/saros/net/xmpp/contact/XMPPContactsService.java

```diff
@@ -0,0 +1,425 @@
+
+    for (String address : addresses) {
+      XMPPContact contact = contacts.get(address);
+      if (contact == null) {
+        log.error("Should not happen: Contact for " + address + " not found!");
+        continue;
+      }
+
+      RosterEntry entry = roster.getEntry(address);
+      if (entry == null) {
+        log.error("Should not happen: Contact found, but RosterEntry is missing for: " + address);
+        continue;
+      }
+
+      if (handlePendingSubscription(contact, entry)) notifyListeners(contact, UpdateType.STATUS);
+
+      if (contacts.setContactGroups(contact, entry.getGroups()))
+        notifyListeners(contact, UpdateType.GROUP_MAPPING);
+
+      if (contact.setNickname(entry.getName()))
+        notifyListeners(contact, UpdateType.NICKNAME_CHANGED);
+    }
+  }
+
+  private void contactResourceChangedPresence(Presence presence) {
+    if (roster == null) return;
+
+    String address = presence.getFrom();
+    XMPPContact contact = contacts.get(address);
+    if (contact == null) {
+      log.error("Should not happen: Contact " + address + " for presence update not found!");
+      return;
+    }
+
+    JID fullJid = new JID(address);
+    if (presence.isAvailable()) {
+      handleContactResourceAvailable(contact, fullJid, presence);
+    } else {
+      handleContactResourceUnavailable(contact, fullJid);
+    }
+  }
+
+  private void contactCanceledSubscription(JID jid) {
+    XMPPContact contact = contacts.get(jid.getRAW());
+    if (contact == null) {
+      log.debug("Strange behaviour: no contact found for canceled subscription from " + jid);
+      return;
+    }
+
+    discoveryService.removeResources(contact.getResources());
+    contact.removeResources();
+    contact.setBaseStatus(ContactStatus.TYPE_SUBSCRIPTION_CANCELED);
+    notifyListeners(contact, UpdateType.STATUS);
+  }
+
+  private boolean handlePendingSubscription(XMPPContact contact, RosterEntry entry) {
+    if (entry.getStatus() == ItemStatus.SUBSCRIPTION_PENDING)
+      return contact.setBaseStatus(ContactStatus.TYPE_SUBSCRIPTION_PENDING);
+    return false;
+  }
+
+  private void handleContactResourceAvailable(XMPPContact contact, JID fullJid, Presence presence) {
+    contact.setSubscribed();
+
+    ContactStatus status;
+    if (presence.isAway()) {
+      status = ContactStatus.newInstance(Type.AWAY, presence.getMode().toString());
+    } else {
+      status = ContactStatus.TYPE_AVAILABLE;
+    }
+    if (contact.setResourceStatus(fullJid, status)) notifyListeners(contact, UpdateType.STATUS);
+
+    discoveryService.querySarosSupport(fullJid, handleQuerySarosSupportResult(fullJid, contact));
+  }
+
+  private void handleContactResourceUnavailable(XMPPContact contact, JID fullJid) {
+    discoveryService.removeResources(Collections.singletonList(fullJid));
+    if (contact.removeResource(fullJid)) notifyListeners(contact, UpdateType.STATUS);
+  }
+
+  /**
+   * Consumer to be called by DiscoveryService after a positive query.
```

Comment:
This is still confusing. Basically this method creates a consumer / callback - more specific a closure - for the given arguments. Can you please rename the method ? It suggests it would handle something which is not quite what it does. It generates something that handle something.

Reply:
How about `createFeatureQueryResultHandler`?

Reply:
Seems fine.

Reply:
done